cdc/puller/unified_sorter.go

```diff
@@ -0,0 +1,832 @@
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case out <- event:
+				}
+			}
+			counter += 1
+
+			if event.CRTs < task.lastTs {
+				log.Fatal("unified sorter: ts regressed in one backEnd, bug?", zap.Uint64("cur-ts", event.CRTs), zap.Uint64("last-ts", task.lastTs))
+			}
+			task.lastTs = event.CRTs
+
+			select {
+			case <-resolvedTicker.C:
+				err := sendResolvedEvent(event.CRTs - 1)
+				if err != nil {
+					return errors.Trace(err)
+				}
+			default:
+			}
+
+			event, err := task.backend.readNext()
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			if event == nil {
+				// EOF
+				delete(workingSet, task)
+				delete(pendingSet, task)
+
+				err := task.dealloc()
+				if err != nil {
+					return errors.Trace(err)
+				}
+
+				continue
+			}
+
+			if event.CRTs > minResolvedTs || (event.CRTs == minResolvedTs && event.RawKV.OpType == model.OpTypeResolved) {
+				// we have processed all events from this task that need to be processed in this merge
+				if event.CRTs > minResolvedTs || event.RawKV.OpType != model.OpTypeResolved {
+					pendingSet[task] = event
+				}
+				err := retire(task)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				continue
+			}
+
+			if counter%10000 == 0 {
+				log.Debug("Merging progress", zap.Int("counter", counter))
+			}
+			heap.Push(sortHeap, &sortItem{
+				entry: event,
+				data:  task,
+			})
+		}
+
+		if len(workingSet) != 0 {
+			log.Fatal("unified sorter: merging ended prematurely, bug?", zap.Uint64("resolvedTs", minResolvedTs))
+		}
+
+		if counter >= 10000 {
+			log.Debug("Unified Sorter: merging ended", zap.Uint64("resolvedTs", minResolvedTs), zap.Int("count", counter))
+		}
+		err := sendResolvedEvent(minResolvedTs)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		return nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case task := <-in:
+			if task == nil {
+				return errors.New("Unified Sorter: nil flushTask, exiting")
+			}
+
+			if task.backend != nil {
+				pendingSet[task] = nil
+			} // otherwise it is an empty flush
+
+			if lastResolvedTs[task.heapSorterID] < task.maxResolvedTs {
+				lastResolvedTs[task.heapSorterID] = task.maxResolvedTs
+			}
+
+			minTemp := uint64(math.MaxUint64)
+			for _, ts := range lastResolvedTs {
+				if minTemp > ts {
+					minTemp = ts
+				}
+			}
+
+			if minTemp > minResolvedTs {
+				minResolvedTs = minTemp
+				err := onMinResolvedTsUpdate()
+				if err != nil {
+					return errors.Trace(err)
+				}
+			}
+		}
+	}
+}
+
+var (
+	pool   *backEndPool
+	poolMu sync.Mutex
+)
+
+// UnifiedSorter provides both sorting in memory and in file. Memory pressure is used to determine which one to use.
+type UnifiedSorter struct {
+	inputCh  chan *model.PolymorphicEvent
+	outputCh chan *model.PolymorphicEvent
+	dir      string
+	pool     *backEndPool
+}
+
+// NewUnifiedSorter creates a new UnifiedSorter
+func NewUnifiedSorter(dir string) *UnifiedSorter {
```

Comment:
Please create a file "LOCK" and lock it to prevent multiple changefeeds use the same dir.

Reply:
The `backEndPool`, which manages the allocation of files, is shared by all changefeeds in the same CDC process, and all files are prefixed by the PID.